bridges/snowbridge/pallets/ethereum-client/src/migration.rs

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+use super::*;
+use frame_support::traits::OnRuntimeUpgrade;
+use log;
+use frame_support::traits::PalletInfoAccess;
+use frame_support::migration::clear_storage_prefix;
+use frame_support::migration::storage_iter;
+
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+
+pub mod v1 {
+    use frame_support::{pallet_prelude::*, weights::Weight};
+    use frame_support::storage::storage_prefix;
+
+    use super::*;
+
+    const LOG_TARGET: &str = "ethereum-client::migration";
+
+    pub struct ExecutionHeaderCleanupOnUpgrade<T>(
+        sp_std::marker::PhantomData<T>,
+    );
+    impl<T> OnRuntimeUpgrade
+    for ExecutionHeaderCleanupOnUpgrade<T>
+        where
+            T: Config,
+    {
+        fn on_runtime_upgrade() -> Weight {
+
+            // To delete
+            // LatestExecutionState
+            // ExecutionHeaders
+            // ExecutionHeaderIndex
+            // ExecutionHeaderMapping
+            log::info!(target: LOG_TARGET, "Running migration v1.");
+            if sp_io::storage::get(&LatestExecutionState::<T>::hashed_key()).is_some() {
+                LatestExecutionState::<T>::kill();
```

Comment:
@vgeddes @yrong I added the execution header storage items back in this PR, because it is much easier and less complicated than having to deal with raw storage keys to delete storage items ([I started off doing with the raw keys](https://github.com/Snowfork/polkadot-sdk/tree/execution-header-migration) but ran into some issues removing the map items one by one) . We can remove the execution header storage items once the migration has run successfully. 

Let me know if you have any concerns with this approach in general.

It is still WIP but just wanted to get your feedback on this before I continue.

Reply:
I'd prefer the raw keys way rather than adding these deprecated/removed primitives back.
 
Btw:  I would suggest to remove `ExecutionHeaders` in the `Multi-Block` manner or the chain could get stuck. There is a PR  https://github.com/paritytech/polkadot-sdk/pull/1781 for that FYI.

Reply:
Clara, what issues did you run into? Surely `clear_storage_prefix` doesn't require any raw key handling, looking at how its used?

Reply:
According to other example in the Polkadot-SDK, should be able to remove all keys in the `ExecutionHeaders` storage map by running:
```
clear_storage_prefix(
			<Pallet<T, I>>::name().as_bytes(),
			b"ExecutionHeaders",
			b"",
			None,
			None,
		)
```

Reply:
Yip, I tested that out: https://github.com/Snowfork/polkadot-sdk/pull/131/files#diff-8e67b18e3a9852ee4f6ce35f1c196806813e216cd23e8a9a71725c1e077f763bR58 I tried the `maybe_limit` param to limit the number of headers deleted in a block (since the map could be large and potentially not all deleted in one go), but that didn't work. Perhaps I misunderstood and the function deletes all the headers it can and returns a cursor to the next indice for the next block. I'll go though the code

Reply:
Maybe I am overthinking it and it can easily be deleted in a single block.

Reply:
My concern here is that we configured the limit as 8192*20 which definitely can't be deleted in a single block.

Reply:
TBH I'm not sure much was gained from this migration, considering the complexity of multi-block migration(MBM) could we just leave the old storage there without migrating?

Reply:
I think its good practice to clean it up, right?

Imo, keeping the structs is an easier way to do the cleanup and it can be done fairly simply, with little downside. We can delete a few headers every block and complete the cleanup, and then just remove the leftover code in the next release.

Reply:
I have looked into `clear_storage_prefix`. In my test, the delete `limit` had no effect because [all the in-memory (overlay) changes are deleted](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/state-machine/src/ext.rs#L460).  All the headers in my tests were in the overlay, not the backend.

The deletion of storage items already committed to the [backend will be limited to the](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/state-machine/src/ext.rs#L461) `maybe_limit` provided.

Using `clear_storage_prefix` to clear the storage will require a new storage item anyway, since a [cursor to the next items to be deleted in the backend should be stored](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/support/src/storage/unhashed.rs#L122). Imo, then we could just as well use the existing storage as is and deleted the items through the storage item interfaces. I think we are overcomplicating it and it doesn't really matter for such a temporary feature.

In any case, I'll look at the Multi-block migration too.

Reply:
Or maybe we can just add a temp extrinsic only for the deletion(called by anyone anytime), in the extrinsic we config a limit(e.g. 100) as max entries which is safe for a single block?

Reply:
How would it be better to have an extrinsic than a migration? You could do the same in the migration, and it has the added benefit that it will be invoked automatically, no one needs to call it.

Reply:
Yeah, the extrinsic is not ideal either. The main idea here is to ensure a heavy migration job is executed in multiple blocks and does not block the chain. There are different ways to achieve that, another example from [pallet-xcm](https://github.com/paritytech/polkadot-sdk/blob/8342947b8e22d398ab981d33d9d48756a3160f77/polkadot/xcm/pallet-xcm/src/lib.rs#L780-L782) is to do that on `on_initialize` phase FYI.

Another concern is that we add the previous storage/primitives back for temp usage. There is a macro https://paritytech.github.io/polkadot-sdk/master/frame_support/attr.storage_alias.html we can add the legacy storage/primitives only in the migration module without polluting the current codes.

Reply:
> Yeah, the extrinsic is not ideal either. The main idea here is to ensure a heavy migration job is executed in multiple blocks and does not block the chain.

Yes, but you can do it using the storage interfaces as well. In this part I create an iterator on `ExecutionHeaders` and simply delete the next item: https://github.com/Snowfork/polkadot-sdk/pull/130/files#diff-8e67b18e3a9852ee4f6ce35f1c196806813e216cd23e8a9a71725c1e077f763bR48-R58 I think this is much more readable and clear.

> Another concern is that we add the previous storage/primitives back for temp usage. There is a macro https://paritytech.github.io/polkadot-sdk/master/frame_support/attr.storage_alias.html I'd suggest if necessary only adding these storage/primitives in the migration module without polluting the current codes.

Would benefit would using the `storage_alias` have above using the normal `storage`?

Reply:
> Would benefit would using the storage_alias have above using the normal storage?

It seems that with `storage_alias` the types are not added to the external pallet API, and are only visible within the migration code. So this is exactly what we want. It is much more ergonomic than `clear_storage_prefix`.

> vgeddes your thoughts here would also be appreciate, so I can perhaps work on this tomorrow.

Lets temporarily halt this work until BridgeHub has been upgraded with the built-in MBM support from https://github.com/paritytech/polkadot-sdk/pull/1781.

Then implementing this migration is going to be much simpler and safer than doing a multi block migration manually.

> Looks like MBM is appropriate for major chain upgrades like core-time. We can't block the chain until our migration has completed.

I looked through `pallet-migrations` in FRAME, and does look like it can be used for any general purpose migration.

Reply:
Please add a ticket though to track this work

Reply:
Updated ticket [SNO-947 Add storage migration in beacon client to remove obsolete storage fields](https://linear.app/snowfork/issue/SNO-947) with more details.